### ISSUE TITLE
Width multiplier

### ISSUE THREAD

**Issue by Skyfrei** [google-research/morph-net#219]
Hi I am currently ending up with a shrunk network which is all good, but then when i am expanding this is happening, meaning my network is ending up larger than what it started as initially.

```py
        self.conv_layer1 = nn.Conv2d(in_channels=3, out_channels=64, kernel_size=kernel, padding = 1)
        self.relu = nn.ReLU(inplace=True)
        self.conv_layer2 = nn.Conv2d(in_channels=64, out_channels=64, kernel_size = kernel)
        self.conv_layer3 = nn.Conv2d(in_channels=64, out_channels=64, kernel_size=kernel)
        self.pool1 = nn.MaxPool2d(kernel_size = kernel, stride = kernel)
        self.conv_layer4 = nn.Conv2d(in_channels=64,  out_channels=128, kernel_size=kernel)
        self.conv_layer5 = nn.Conv2d(in_channels=128,  out_channels=128, kernel_size=kernel)
        self.pool2 = nn.MaxPool2d(kernel_size=kernel, stride = kernel) # first
        self.conv_layer6 = nn.Conv2d(in_channels=128,  out_channels=256, kernel_size=kernel-1, padding = 1)
        self.conv_layer7 = nn.Conv2d(in_channels=256,  out_channels=256, kernel_size=kernel-1)
        self.conv_layer8 = nn.Conv2d(in_channels=256,  out_channels=512, kernel_size=kernel)
        self.conv_layer9 = nn.Conv2d(in_channels=512,  out_channels=1024, kernel_size=kernel)
        self.pool2 = nn.AvgPool2d(kernel_size = kernel, stride = kernel) # second
        self.fc1 = nn.Linear(1024 * 3 * 3, num_classes)
```

```py
  def expand_network(self, real_data, max_flops):
        reached = False
        init_modules = copy.deepcopy([mod for _, mod in self.named_modules() if isinstance(mod, nn.Conv2d)])

        while reached is False:
            j = 0
            for name, module in self.named_modules():
                if isinstance(module, nn.Conv2d):
                    out_channels = int(init_modules[j].out_channels * self.omega)
                    in_channels = 3
                    if module.in_channels != 3:
                        in_channels = int(init_modules[j].in_channels * self.omega)
                    module.out_channels = out_channels
                    module.in_channels = in_channels
                    if j == 0:
                        print(module.out_channels)
                    j = j + 1

            if self.calculate_conv_flops2(real_data) >= max_flops:
                reached = True
                for _, module in self.named_modules():
                    if isinstance(module, nn.Conv2d):
                        print(module)
            else:
                self.omega = self.omega + 0.01
        print(f"Omega converged at : {self.omega}")

```

how i calculate flop:

```py

 def calculate_conv_flops2(self, real_data_point):
        module_input = real_data_point.to(device)
        total_regularizer = 0
        activations = {}

        input_activations = {}
        output_activations = {}

        def hook_fn(module, input, output):
            input_activations[module] = input[0].detach()  # input is a tuple
            output_activations[module] = output.detach()
        hooks = []

        for name, module in self.named_modules():
            if isinstance(module, nn.Conv2d):
                hooks.append(module.register_forward_hook(hook_fn))
        _ = self.forward(module_input)

        for name, module in self.named_modules():
            if isinstance(module, nn.Conv2d):
                out_h = module.weight.size(2)
                out_w = module.weight.size(3)
                f = module.in_channels
                g = module.out_channels
                C = 2 * out_h * out_w * f * g     + out_h * out_w * g

                in_act = input_activations[module]
                # Check which input neurons (channels) are alive
                A = (in_act.abs().sum(dim=[0, 2, 3]) > 0).float().sum()  # [in_channels]
                out_act = output_activations[module]
                # Check which output neurons (channels) are alive
                B = (out_act.abs().sum(dim=[0, 2, 3]) > 0).float().sum()  # [out_channels]
                total_regularizer += C * A * B

        for hook in hooks:
            hook.remove()
        return total_regularizer
```

Conv2d(3, 217, kernel_size=(2, 2), stride=(1, 1), padding=(1, 1))
Conv2d(217, 81, kernel_size=(2, 2), stride=(1, 1))
Conv2d(81, 81, kernel_size=(2, 2), stride=(1, 1))
Conv2d(81, 217, kernel_size=(2, 2), stride=(1, 1))
Conv2d(217, 217, kernel_size=(2, 2), stride=(1, 1))
Conv2d(217, 435, kernel_size=(1, 1), stride=(1, 1), padding=(1, 1))
Conv2d(435, 407, kernel_size=(1, 1), stride=(1, 1))
Conv2d(407, 843, kernel_size=(2, 2), stride=(1, 1))
Conv2d(843, 1740, kernel_size=(2, 2), stride=(1, 1))
Omega converged at : 3.399999999999971

Is this a desired behaviour because I think making the new network be bigger than the initial network loses the point or am i missing something?

**Comment by Skyfrei** [comment]
Issue fixed, 

I was just pruning so much and masking so many neurons that when expanding omega was blowing up.
